opencog/attention/experimental/HebbianUpdatingAgent.cc

```diff
@@ -52,7 +57,12 @@ void HebbianUpdatingAgent::run()
     std::back_insert_iterator< std::vector<Handle> > out_hi(atoms);
 
     _as->get_handle_set_in_attentional_focus(out_hi);
-
+     /*Without adding this sleep code right below the above method call,
+     * nlp-parse evaluation thread waits for minutes before it gets a chance to
+     * run.
+     */
+    std::this_thread::sleep_for(std::chrono::nanoseconds(1));
```

Comment:
The way in which it gets slowed is described in detail, at the bottom of bug #2333 -- basically, the agents are causing a huge amount of lock contention -- they are monopolizing the lock on the atomspace, because `_as->get_handle_set_in_attentional_focus` takes a long time to run, and is called over and over.  Pretty much no one else gets a chance to get in there and do anything.

The reason that the nano-sleep/thread_yeild works is because either of these calls descend to the Linux operating system kernel, and tell the process/thread scheduler that the current thread should be suspended, and that some other thread/process should be given a chance to run. The kernel then searches for runnable threads -- and finds one -- one of the other threads, that is waiting for the atomspace lock. Its scheduled, and it runs, and does it's stuff.

Without the nanosleep/yeild, that other thread almost never gets a chance to run: this thread is constantly releasing and grabbing the lock, and so, statistically, the chances are very small that the linux kernel scheduler will run, and find the lock not held.  Thus, the other thread is "starved".  https://en.wikipedia.org/wiki/Starvation_(computer_science)